alonzo/impl/src/Cardano/Ledger/Alonzo/PlutusScriptApi.hs

```diff
@@ -119,6 +128,81 @@ collectNNScriptInputs pp tx utxo =
           cost <- maybeToList (Map.lookup PlutusV1 (getField @"_costmdls" pp))
       ]
 
+-- ========================================================================
+
+-- | When collecting inputs for twophase scripts, 3 things can go wrong.
+data CollectError crypto
```

Comment:
I wonder if we could give this a name to make it clear that this is a logic error or sorts (ie we don't ever expect to see these, and if we do, we know there is a bug). `CollectLogicError`?